server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -22,6 +24,50 @@
 
 public abstract class Mapper implements ToXContentFragment, Iterable<Mapper> {
 
+    public static final NodeFeature STORE_SOURCE_MAPPER_PARAM = new NodeFeature("mapper.source.store_param");
+
+    static final String STORE_SOURCE_PARAM = "store_source";
+
+    // Only relevant for synthetic source mode.
+    public enum StoreSourceMode {
+        NONE("none"),      // No source recording
+        ARRAYS("arrays"),  // Store source for arrays of mapped fields
+        ALL("all");        // Store source for both singletons and arrays of mapped fields
+
+        StoreSourceMode(String name) {
+            this.name = name;
+        }
+
+        static StoreSourceMode fromString(String input) {
+            if (input.equals(NONE.name)) {
+                return NONE;
+            }
+            if (input.equals(ALL.name)) {
+                return ALL;
+            }
+            if (input.equals(ARRAYS.name)) {
+                return ARRAYS;
+            }
+            throw new IllegalArgumentException("Unknown " + STORE_SOURCE_PARAM + " value [" + input + "]");
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+
+        private final String name;
+    }
+
+    // Setting StoreSourceMode to ENABLED at the index level is equivalent to disabling synthetic source, which is not desired.
+    // Since the only valid option is to track array source by default, we use a boolean index setting for it.
+    public static final Setting<Boolean> STORE_ARRAY_SOURCE_SETTING = Setting.boolSetting(
+        "index.mapping.store_array_source",
+        false,
+        Setting.Property.IndexScope,
+        Setting.Property.ServerlessPublic
+    );
+
```

Comment:
Can we restrict this feature to the `ObjectMapper` only? `Mapper` is too generic to define this option.

Reply:
We actually need this mostly for leaf fields - cases like args.

Reply:
ah that's what I misunderstood, it's for all arrays not only objects, thanks.